### PR TITLE
CASM-4937 Phase 1: allow globs in version specs

### PR DESCRIPTION
**Pull request by mtupitsyn**
## Summary and Scope

1. Allow globs (`*`) in version specs for:
* assets `ARTIFACT_ID`'s and `KERNEL_VERSION`
* RPMs in `rpm/**/*.yaml`
* container images in `docker/index.yaml`
* Helm charts in `manifests/*.yaml`
2. Write actual versions (with globs resolved) into a file and publish this file along with CSM build

## Issues and Related PRs

Phase 1 of CASM-4937:
* Resolves CASM-4938
* Resolves CASM-4939
* Resolves CASM-4940
* Resolves CASM-4941
* Resolves CASM-4942

### Tested on:

  * Local development environment
  * Jenkins

### Test description:

  * Added sample globs to manifests
  * Ensured that versions are resolved properly and published in `csm-<RELEASE_VERSION>-versions.yaml` file: https://jenkins.algol60.net/job/Cray-HPE/job/csm/job/feature%2Fversion-globs/8/artifact/dist/csm-1.6.1-alpha.5-6-g087ba084-versions.yaml/*view*/

## Risks and Mitigations

Low - build time change only

### CODE REVIEW

File: assets.sh [hunk 1]
```diff
@@ -31,47 +34,61 @@ NCN_ARCH='x86_64'
...
 # All images must use the same, exact kernel version.
-KERNEL_VERSION='6.4.0-150600.23.17-default'
-# NOTE: The kernel-default-debuginfo package version needs to be aligned
-# to the KERNEL_VERSION. Always verify and update the correct version of
-# the kernel-default-debuginfo package when changing the KERNEL_VERSION
-# by doing a zypper search for the corresponding kernel-default-debuginfo package
-# in the SLE-Module-Basesystem update_debug repo
-# zypper --plus-repo=https://${ARTIFACTORY_USER}:${ARTIFACTORY_TOKEN}@artifactory.algol60.net/artifactory/sles-mirror/Updates/SLE-Module-Basesystem/15-SP4/x86_64/update_debug se -s kernel-default-debuginfo
-KERNEL_DEFAULT_DEBUGINFO_VERSION="${KERNEL_VERSION//-default/}.1"
+KERNEL_VERSION='6.4.0-*-default'
```
Comment (mtupitsyn) [inline assets.sh]: Temporary change to test glob processing. Represents a sample change which we may have in integration branch. Will be reverted before merge into release branch.

File: assets.sh [hunk 2]
```diff
@@ -31,47 +34,61 @@ NCN_ARCH='x86_64'
...
 
 # The image ID may not always match the other images and should be defined individually.
-KUBERNETES_IMAGE_ID=6.2.33
+KUBERNETES_IMAGE_ID=6.2.*
```
Comment (mtupitsyn) [inline assets.sh]: Temporary change to test glob processing. Represents a sample change which we may have in integration branch. Will be reverted before merge into release branch.

File: assets.sh [hunk 3]
```diff
@@ -31,47 +34,61 @@ NCN_ARCH='x86_64'
...
+KERNEL_VERSION=${KERNEL_VERSION%-${KUBERNETES_IMAGE_ID}-${NCN_ARCH}.kernel}
 KUBERNETES_ASSETS=(
     "https://artifactory.algol60.net/artifactory/csm-images/stable/kubernetes/${KUBERNETES_IMAGE_ID}/kubernetes-${KUBERNETES_IMAGE_ID}-${NCN_ARCH}.squashfs"
     "https://artifactory.algol60.net/artifactory/csm-images/stable/kubernetes/${KUBERNETES_IMAGE_ID}/${KERNEL_VERSION}-${KUBERNETES_IMAGE_ID}-${NCN_ARCH}.kernel"
     "https://artifactory.algol60.net/artifactory/csm-images/stable/kubernetes/${KUBERNETES_IMAGE_ID}/initrd.img-${KUBERNETES_IMAGE_ID}-${NCN_ARCH}.xz"
 )
 
 # The image ID may not always match the other images and should be defined individually.
-PIT_IMAGE_ID=6.2.33
+PIT_IMAGE_ID=6.2.*
```
Comment (mtupitsyn) [inline assets.sh]: Temporary change to test glob processing. Represents a sample change which we may have in integration branch. Will be reverted before merge into release branch.

File: assets.sh [hunk 4]
```diff
@@ -31,47 +34,61 @@ NCN_ARCH='x86_64'
...
+# Resolve globs in PIT_IMAGE_ID, e.g. 6.2.* > 6.2.30
+PIT_IMAGE_ID=$(basename $(dirname $(resolve_globs "csm-images" "stable/pre-install-toolkit/${PIT_IMAGE_ID}" "pre-install-toolkit-${PIT_IMAGE_ID}-${NCN_ARCH}.iso")))
 PIT_ASSETS=(
     "https://artifactory.algol60.net/artifactory/csm-images/stable/pre-install-toolkit/${PIT_IMAGE_ID}/pre-install-toolkit-${PIT_IMAGE_ID}-${NCN_ARCH}.iso"
 )
 
 # The image ID may not always match the other images and should be defined individually.
-STORAGE_CEPH_IMAGE_ID=6.2.33
+STORAGE_CEPH_IMAGE_ID=6.2.*
```
Comment (mtupitsyn) [inline assets.sh]: Temporary change to test glob processing. Represents a sample change which we may have in integration branch. Will be reverted before merge into release branch.

File: assets.sh [hunk 5]
```diff
@@ -31,47 +34,61 @@ NCN_ARCH='x86_64'
...
+STORAGE_CEPH_IMAGE_ID=$(basename $(dirname $(resolve_globs "csm-images" "stable/storage-ceph/${STORAGE_CEPH_IMAGE_ID}" "storage-ceph-${STORAGE_CEPH_IMAGE_ID}-${NCN_ARCH}.squashfs")))
 STORAGE_CEPH_ASSETS=(
     "https://artifactory.algol60.net/artifactory/csm-images/stable/storage-ceph/${STORAGE_CEPH_IMAGE_ID}/storage-ceph-${STORAGE_CEPH_IMAGE_ID}-${NCN_ARCH}.squashfs"
     "https://artifactory.algol60.net/artifactory/csm-images/stable/storage-ceph/${STORAGE_CEPH_IMAGE_ID}/${KERNEL_VERSION}-${STORAGE_CEPH_IMAGE_ID}-${NCN_ARCH}.kernel"
     "https://artifactory.algol60.net/artifactory/csm-images/stable/storage-ceph/${STORAGE_CEPH_IMAGE_ID}/initrd.img-${STORAGE_CEPH_IMAGE_ID}-${NCN_ARCH}.xz"
 )
 
 # The image ID may not always match the other images and should be defined individually.
-COMPUTE_IMAGE_ID=6.2.33
+COMPUTE_IMAGE_ID=6.2.*
```
Comment (mtupitsyn) [inline assets.sh]: Temporary change to test glob processing. Represents a sample change which we may have in integration branch. Will be reverted before merge into release branch.

File: docker/index.yaml [hunk 6]
```diff
@@ -26,7 +26,7 @@ artifactory.algol60.net/csm-docker/stable:
 
     # cray-sat is not included in any Helm charts
     cray-sat:
-      - 3.33.1
+      - 3.33.*
```
Comment (mtupitsyn) [inline docker/index.yaml]: Temporary change to test glob processing. Represents a sample change which we may have in integration branch. Will be reverted before merge into release branch.

File: manifests/nexus.yaml [hunk 7]
```diff
@@ -33,5 +33,5 @@ spec:
   charts:
   - name: cray-nexus
     source: csm-algol60
-    version: 0.12.2
+    version: 0.12.*
```
Comment (mtupitsyn) [inline manifests/nexus.yaml]: Temporary change to test glob processing. Represents a sample change which we may have in integration branch. Will be reverted before merge into release branch.

File: rpm/cray/csm/noos/index.yaml [hunk 8]
```diff
@@ -44,8 +44,8 @@ https://artifactory.algol60.net/artifactory/csm-rpms/hpe/stable/noos/:
     - csm-node-identity-1.0.22-1.noarch
     - csm-ssh-keys-1.6.3-1.noarch
     - csm-ssh-keys-roles-1.6.3-1.noarch
-    - csm-testing-1.18.9-1.noarch
-    - goss-servers-1.18.9-1.noarch
+    - csm-testing-1.18.*-1.noarch
+    - goss-servers-1.18.*-1.noarch
```
Comment (mtupitsyn) [inline rpm/cray/csm/noos/index.yaml]: Temporary change to test glob processing. Represents a sample change which we may have in integration branch. Will be reverted before merge into release branch.

**Review by studenym-hpe — APPROVED** [review]

**Pull request merged**